### PR TITLE
Upgrade ginkgo, gomega and ginkgolinter

### PR DESCRIPTION
**Pull request by nunnatsa**
Upgrade ginkgo to v2.9.4
Upgrade gomega to v1.27.6
upgrade ginkgolinter to v0.12.0 and fix the new findings.

```release-note
None
```

### CODE REVIEW

**Comment by enp0s3** [comment]
@nunnatsa Thank you!
/approve

**Comment by enp0s3** [comment]
/hold
I see some compilation errors, can you please check?

File: tests/storage/snapshot.go [hunk 1]
```diff
@@ -1053,7 +1053,7 @@ var _ = SIGDescribe("VirtualMachineSnapshot Tests", func() {
 
 				errStr := fmt.Sprintf("VolumeSnapshots (%s) missing", *vb.VolumeSnapshotName)
 				Expect(snapshot.Status.Error).ToNot(BeNil())
-				Expect(snapshot.Status.Error.Message).ToNot(Equal(errStr))
+				Expect(snapshot.Status.Error.Message).ToNot(HaveValue(Equal(errStr)))
```
Comment (nunnatsa) [inline tests/storage/snapshot.go:1056]: This is now failing in e2e.

The reason it was passed until now is that `snapshot.Status.Error.Message` is a pointer, compared to a string. This is always true.

With the fix, it seems that the value actually does equal to the value of the `errStr` variable.

@mhenriks - could you please take a look? How should we address it? Should we revert the assertion logic?

File: tests/vm_test.go [hunk 2]
```diff
@@ -1282,7 +1282,7 @@ var _ = Describe("[rfe_id:1177][crit:medium][vendor:cnv-qe@redhat.com][level:com
 							return false
 						}, 10*time.Second, 1*time.Second).Should(BeTrue())
 					} else {
-						Eventually(Expect(launcherPod.Status.Phase).To(Equal(k8sv1.PodFailed)), 160*time.Second, 1*time.Second).Should(BeTrue())
+						Eventually(Expect(launcherPod.Status.Phase).To, 160*time.Second, 1*time.Second).WithArguments(Equal(k8sv1.PodFailed)).Should(BeTrue())
```
Comment (akalenyu) [inline tests/vm_test.go:1285]: Is this a common practice? (Expect inside an eventually)

**Review by akalenyu — COMMENTED** [review]
Just a question

File: tests/storage/snapshot.go (hunk 1 above)
Comment (akalenyu) [reply tests/storage/snapshot.go]: Yeah I believe the original intention was that the error is == errStr, mhenriks will keep me honest

File: tests/storage/reservation.go [hunk 3]
```diff
@@ -234,9 +234,11 @@ var _ = SIGDescribe("[Serial]SCSI persistent reservation", Serial, func() {
 				"there are NO registered reservation keys")).To(BeTrue())
 			Expect(checkResultCommand(vmi, "sg_persist -o -G  --param-sark=12345678 /dev/sda",
 				"Peripheral device type: disk")).To(BeTrue())
-			Eventually(checkResultCommand(vmi, "sg_persist -i -k /dev/sda",
-				"1 registered reservation key follows:\r\n    0x12345678\r\n"),
-				60*time.Second, 10*time.Second).Should(BeTrue())
+			Eventually(checkResultCommand,
+
```
Comment (rmohr) [inline tests/storage/reservation.go:238]: Is this empty line an artifact of some auto-formatting or a manual thing?

**Review by rmohr — COMMENTED** [review]
the arm64 lane fails on something unrelated.

I am not sure I am a fan of the `WithArguments` exra verbosity, but given that we do it now in a consistent fashion makes it +1 from me.

One question before lgtm.

File: tests/vm_test.go (hunk 2 above)
Comment (akalenyu) [reply tests/vm_test.go:1285]: Quick github search says it's not very common (mostly kubevirt forks):
https://github.com/search?q=%22Eventually%28Expect%28%22&type=code

File: tests/storage/reservation.go (hunk 3 above)
Comment (nunnatsa) [reply tests/storage/reservation.go:238]: Some kind of auto formatting. Will fix.

File: tests/storage/reservation.go [hunk 4]
```diff
@@ -262,9 +263,11 @@ var _ = SIGDescribe("[Serial]SCSI persistent reservation", Serial, func() {
...
```
Comment (rmohr) [inline tests/storage/reservation.go:267]: another one

File: tests/vm_test.go (hunk 2 above)
Comment (akalenyu) [reply tests/vm_test.go]: I would just return the `launcherPod.Status.Phase` and compare it to the desired PodPhase

**Comment by nunnatsa** [comment]
> I am not sure I am a fan of the `WithArguments` exra verbosity

I also think it's can be confusing.

The alternative is something like:

```go
Eventually(func() bool {
		return checkResultCommand(vmi, "sg_persist -i -k /dev/sda", "1 registered reservation key follows:\r\n    0x12345678\r\n")
	}).Within(60*time.Second).
	WithPolling(10*time.Second).
	Should(BeTrue())
```
Or
```go
Eventually(func(g Gomega) {
		g.Expect(checkResultCommand(vmi, "sg_persist -i -k /dev/sda", "1 registered reservation key follows:\r\n    0x12345678\r\n")).Should(BeTrue())
	}).Within(60*time.Second).
	WithPolling(10*time.Second).
	Should(Succeed())
```

But this is much harder to implement in the linter.

**Comment by rmohr** [comment]
> > I am not sure I am a fan of the `WithArguments` exra verbosity
> 
> I also think it's can be confusing.
> 
> The alternative is something like:
> 
> ```go
> Eventually(func() bool {
> 		return checkResultCommand(vmi, "sg_persist -i -k /dev/sda", "1 registered reservation key follows:\r\n    0x12345678\r\n")
> 	}).Within(60*time.Second).
> 	WithPolling(10*time.Second).
> 	Should(BeTrue())
> ```
> 
> Or
> 
> ```go
> Eventually(func(g Gomega) {
> 		g.Expect(checkResultCommand(vmi, "sg_persist -i -k /dev/sda", "1 registered reservation key follows:\r\n    0x12345678\r\n")).Should(BeTrue())
> 	}).Within(60*time.Second).
> 	WithPolling(10*time.Second).
> 	Should(Succeed())
> ```
> 
> But this is much harder to implement in the linter.
> 
> I think that linters in general, should provide a working solution, and we always can review and customize it.

For as long as we can still use the patterns above (which I think we can), streamlining the usage like you do right now works for me.

**Comment by rmohr** [comment]
/lgtm


/hold

Adding a hold for @akalenyu. Feel free to remove when ready.

**Comment by akalenyu** [comment]
> /lgtm
> 
> /hold
> 
> Adding a hold for @akalenyu. Feel free to remove when ready.

Yeah will do just that discussion in https://github.com/kubevirt/kubevirt/pull/9712#discussion_r1187225049

File: tests/vm_test.go (hunk 2 above)
Comment (nunnatsa) [reply tests/vm_test.go]: OK @akalenyu, done. But I'll need LGTM again (-;

**Comment by nunnatsa** [comment]
> Yeah will do just that discussion in [#9712 (comment)](https://github.com/kubevirt/kubevirt/pull/9712#discussion_r1187225049)

Done

**Pull request merged**